File: ch3/stack_of_plates/python/Stacks.py
class Stacks:
    def __init__(self, stackSize):
        self.stacks = [[]]
        self.stackSize = stackSize

    def push(self, n):
        if(len(self.stacks[len(self.stacks) - 1]) == self.stackSize):
            self.stacks.append([])
            self.stacks[len(self.stacks) - 1].append(n)
        else:
            self.stacks[len(self.stacks) - 1].append(n)

    def pop(self):
        if(len(self.stacks[len(self.stacks) - 1]) == 0):
            self.stacks.pop()
        self.stacks[len(self.stacks) - 1].pop()

    def top(self):
        return self.stacks[len(self.stacks) - 1][len(self.stacks[len(self.stacks) - 1]) - 1]

    def p(self):
        for i in range(len(self.stacks)):
            for j in range(len(self.stacks[i])):
                print(self.stacks[i][j], end=" ")
            print("")
        print("")

def main():
    s = Stacks(5)
    s.push(1)
    s.push(2)
    s.push(3)
    s.push(4)
    s.push(5)
    s.push(6)
    s.push(7)
    s.push(8)
    s.p()
    s.pop()
    s.pop()
    s.pop()
    s.pop()
    s.pop()
    s.pop()
    s.pop()
    s.p()

if __name__ == "__main__":
    main()

File: ch3/stack_of_plates/java/Stacks.java
import java.util.*;
public class Stacks<Obj>{
    private ArrayList<ArrayList<Obj>> d;
    private int height;
    Stacks(int t_height){
        height = t_height;
        d = new ArrayList<ArrayList<Obj>>();
        ArrayList<Obj> first = new ArrayList<Obj>();
        d.add(first);
    }

    public void push(Obj object){
        if(d.get(d.size() - 1).size() == height){
            ArrayList<Obj> newArr = new ArrayList<Obj>();
            d.add(newArr);
            d.get(d.size() - 1).add(object);
        } else {
            d.get(d.size() - 1).add(object);
        }
    }

    public void pop(){
        if(d.get(d.size() - 1).size() == 0){
            d.remove(d.size() - 1);
        }
        d.get(d.size() - 1).remove(d.get(d.size() - 1).size() - 1);
    }

    public Obj top(){
        return d.get(d.size() - 1).get(d.get(d.size() - 1).size() - 1);
    }

    public void p(){
        for(int i = 0; i < d.size(); i++){
            for(int j = 0; j < d.get(i).size(); j++){
                System.out.print(d.get(i).get(j));
                System.out.print(" ");
            }
            System.out.println("");
        }
        System.out.println("");
    }
}
